sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -955,37 +955,51 @@ class GpuDynamicPartitionDataConcurrentWriter(
       subTables.head
     }
 
+    val maxRecordsPerFile = description.maxRecordsPerFile
     withResource(t) { _ =>
-      if (description.maxRecordsPerFile > 0 &&
-          status.writerStatus.recordsInFile + t.getRowCount >= description.maxRecordsPerFile) {
-        // TODO: This does not handle batch splitting to make sure maxRecordsPerFile
-        //  is not exceeded. See: https://github.com/NVIDIA/spark-rapids/issues/6452
-        // write data
-        val batch = GpuColumnVector.from(t, outDataTypes)
-        // the batch is closed in `write`
-        status.writerStatus.outputWriter.write(batch, statsTrackers)
+      val batch = GpuColumnVector.from(t, outDataTypes)
+      if (!needSplitBatch(maxRecordsPerFile, status.writerStatus.recordsInFile, batch.numRows())) {
         statsTrackers.foreach(_.newBatch(batch))
-        status.writerStatus.recordsInFile += t.getRowCount
+        status.writerStatus.recordsInFile += batch.numRows()
+        status.writerStatus.outputWriter.write(batch, statsTrackers)
+      } else {
+        withResource(batch) { batch =>
+          withResource(GpuColumnVector.from(batch)) {table =>
+            val splitIndexes = getSplitIndexes(
+              maxRecordsPerFile,
+              status.writerStatus.recordsInFile,
+              table.getRowCount()
+            )
 
-        // check max file number
-        assert(status.writerStatus.fileCounter < MAX_FILE_COUNTER,
-          s"File counter ${status.writerStatus.fileCounter} is beyond max value $MAX_FILE_COUNTER")
+            val dataTypes = GpuColumnVector.extractTypes(batch)
 
-        // will create a new file, close the former writer
-        status.writerStatus.outputWriter.close()
+            var needNewWriter = status.writerStatus.recordsInFile >= maxRecordsPerFile
+            withResource(table.contiguousSplit(splitIndexes: _*)) {tabs =>
+              tabs.foreach(b => {
+                if (needNewWriter) {
+                  status.writerStatus.fileCounter += 1
+                  assert(status.writerStatus.fileCounter < MAX_FILE_COUNTER,
```

Comment:
BTW, could you help check whether this assertion (false scenario) will cause resource leaks?

Reply:
I will file a follow-up ticket to track this.